frontend/catalyst/utils/jax_extras.py

```diff
@@ -477,3 +576,372 @@ def make_jaxpr_f(*args, **kwargs):
+
+    Returns:
+        List of arguments containing both explicit and implicit arguments
+        OutputType describing the expanded result
+
+    """
+    return infer_output_type(
+        constants, expanded_inputs, results, expansion_strategy, num_implicit_inputs
+    )
+
+
+def collapse(typ: InputType | OutputType, params: List[TracerLike]) -> List[TracerLike]:
+    """Collapse the expanded list of parameters by cutting the implicit dimension variables, as
+    specified by the type of the parameters."""
+    return [t for t, (_, k) in zip(params, typ) if k]
+
+
+def tracer_index(x: TracerLike, ls: List[TracerLike]) -> Optional[int]:
+    """Return the index of Jax tracer `x` in the list of tracers `ls`, or None."""
+    for i, t in enumerate(ls):
+        if x is t:
+            return i
+    return None
+
+
+def out_type_force_outdbidx(
+    out_type: OutputType,
+    input_idx: int,
+    consts: List[TracerLike],
+    inputs: List[TracerLike],
+    outputs: List[TracerLike],
+) -> OutputType:
+    """Convert all references to the specified input argument of a Python or Jaxpr program into the
+    OutDBIdx format. This function is typically needed to remove references to the loop body
+    iterator.
+
+    Args:
+        out_type: Jax OutputType of results of the program
+        input_idx: Index of the input argument that should be converted from InDBIdx to OutDBIdx
+        consts: Constants used in the program
+        inputs: Expanded list of inputs of the program
+        outputs: Expanded list of outputs of the program
+
+    Returns:
+        New OutputType containing no InDBIdx references to the specified argument.
+    """
+    assert len(out_type) == len(outputs), "Outputs must be expanded"
+    x_in_idx = input_idx
+    x_out_idx = tracer_index([*consts, *inputs][x_in_idx], outputs)
+
+    out_type2 = []
+    for i, ((aval, k), t) in enumerate(zip(out_type, outputs)):
+        assert hasattr(t, "aval"), "Outputs are expected to be Jax tracers or Jaxpr variables"
+        if isinstance(aval, DShapedArray):
+            shape2 = []
+            for d in aval.shape:
+                if isinstance(d, InDBIdx) and d.val == x_in_idx:
+                    assert x_out_idx is not None, (
+                        "Target tracer does not exist in the outputs "
+                        "(see force_implicit_indbidx=True)"
+                    )
+                    assert x_out_idx < i, "Target tracer is not available for OutDBIdx"
+                    shape2.append(OutDBIdx(x_out_idx))
+                else:
+                    shape2.append(d)
+            aval2 = aval.update(shape=tuple(shape2))
+        else:
+            aval2 = copy(aval)
+        out_type2.append((aval2, k))
+    return out_type2
+
+
+class DynshapePrimitive(Primitive):
+    """Primitive containing nested Jaxpr programs accepting and returning values with dynamic
+    shapes."""
+
+    def bind(self, *args, **params):
```

Comment:
I am adding the extended comment, but could you please clarify what exactly do you mean by EvalTrace?

The `bind` handler is called by both `primitive.bind` during the Python tracing and `eval_jaxpr` during the Jaxpr interpretation/tracing.

Reply:
The [EvalTrace](https://github.com/google/jax/blob/a74a78250bd864bb9057bab77efd28c262feb5dd/jax/_src/core.py#L929) is a JAXPR interpreter. I think it is the default interpreter of JAXPR and just evaluates (runs) the primitives. You can see it has this behaviour because the `process_primitive` function just calls the primitives `impl` function.
This interpreter is also explained in the [autodidax tutorial](https://jax.readthedocs.io/en/latest/autodidax.html#evaluation-interpreter).

Reply:
Do my comments make sense @grwl? In that this `bind` acts fundamentally different than the `bind` in JAX, by processing the primitive directly instead of letting the trace process the primitive?

Reply:
I see.. Yes, this is a very good question. I think I might have misused the `bind` API of Jax, given its original purpose seen from the docs. I believe that the cleaner approach would patch DynamicJaxprTracer's `process_primitive` rather than the primitive's bind method
I think that Jax team also have weakened their design compared to the one described in the documentation, I am still looking into this.

Reply:
FYI: I have implemented a [draft trace class](https://github.com/PennyLaneAI/catalyst/blob/b365193abec730f4b1f22b3e878ecc1507945408/frontend/catalyst/utils/jax_extras.py#L954) named `DynamicJaxprTraceEx` supporting dynamic shapes. In order to use it, however, I needed to overload more Jax routines, because the default DynamicJaxprTrace is hardcoded in a number of places in the `partial_eval` module which we currently rely on. Unfortunately, some tests fail with C-segfault, so futher investigation is required.. The code is located in the `dynshape-jax-tracer` branch of the catalyst repo.